### PR TITLE
added support for uploading snap packages directly to the Ubuntu Store.

### PR DESCRIPTION
**Pull request by ricardokirkner**
- added login command to fetch credentials necessary to establish an
  authenticated session for the upload process
- added logout command to clear stored credentials
- added upload command to upload a snap package to the store

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -36,6 +36,12 @@
             ['schema/' + x for x in os.listdir('schema')]),
     ],
     test_suite='snapcraft.tests',
+    install_requires=[
+        'click_toolbelt>=0.4.0',
+    ],
+    tests_require=[
+        'click_toolbelt>=0.4.0',
+    ],
```
Comment (kyrofa) [inline setup.py]: Do these need to be added to debian/control as well?

**Comment by kyrofa** [comment]
Please add your new dependencies to the `.travis.yml` file's installation step.

File: snapcraft/commands/upload.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+def _format_snap_name(snap):
+    snap['arch'] = snap['arch'][0] if len(snap['arch']) == 1 else 'multi'
+    return '{name}_{version}_{arch}.snap'.format(**snap)
+
+
+def main(argv=None):
+    """Upload snap package to the Ubuntu Store."""
+
+    # make sure the full lifecycle is executed
+    snap = lifecycle.execute('strip')
```
Comment (sergiusens) [inline snapcraft/commands/upload.py]: you probably want to run the `snap` command here. This produces no snap.

Maybe some code reorg is needed if the snap command becomes common and just calling snap.main() is to cumbersome (or ugly).

**Comment by elopio** [comment]
Thanks a lot Ricardo!
This needs integration tests, maybe just for the successful paths of login, logout and upload. Does this toolbelt have an easy way to patch the ip of the server so we can fake it?
Let me know if you need a hand, or if you have a better idea of how to test this. The integration tests are written here: https://github.com/ubuntu-core/snapcraft/tree/master/integration_tests

It would also be nice to upload a real snap to the staging server, maybe once every night, but that would require us to keep some credentials in jenkins. I'll discuss about it with @fgimenez.

File: integration_tests/test_upload.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        os.chdir(project_dir)
+
+        snap_file_path = 'assemble_1.0_{}.snap'.format(_get_deb_arch())
+        self.assertThat(snap_file_path, FileExists())
+
+        self.assertIn('Snap assemble_1.0_amd64.snap not found. Running snap '
+                      'stage to create it.', output)
+        self.assertIn('Uploading {}'.format(snap_file_path), output)
+        self.assertIn('Package scan completed.', output)
+        self.assertIn('{} upload failed'.format(snap_file_path), output)
```
Comment (elopio) [inline integration_tests/test_upload.py]: Yes, that's what I had in mind.
I'd prefer to put the whole expected message in a var, maybe as a regex if there are things you want to ignore, and then make one assertion for the output, but that's just a nit.
What I think we need now is to check successful scenarios with a fake server.

File: integration_tests/test_upload.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2015 Canonical Ltd
```
Comment (elopio) [inline integration_tests/test_upload.py]: s/2015/2016.

File: integration_tests/test_upload.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    def test_upload(self):
+        project_dir = 'assemble'
+        output = self.run_snapcraft('upload', project_dir)
+        os.chdir(project_dir)
+
+        snap_file_path = 'assemble_1.0_{}.snap'.format(_get_deb_arch())
+        self.assertThat(snap_file_path, FileExists())
+
+        self.assertIn('Snap assemble_1.0_amd64.snap not found. Running snap '
+                      'stage to create it.', output)
```
Comment (sergiusens) [inline integration_tests/test_upload.py]: can you `s/stage to/step to` to avoid confusion with the `stage` step?

File: runtests.sh [hunk 6]
```diff
@@ -45,16 +45,11 @@ parseargs(){
...
-    # we had python3-flake8 provide flake8
-    python3 -m pep8 $SRC_PATHS
-
-    python3 -m pyflakes $SRC_PATHS
+    python3 /usr/bin/flake8 $SRC_PATHS --exclude=snapcraft/vendor
 
     # mccabe in 'warning' mode as we have high complexity
     mccabe_list=
-    for unit in $(find snapcraft -type f -name '*.py')
+    for unit in $(find snapcraft -type f -name '*.py' -not -path 'snapcraft/vendor/*')
```
Comment (sergiusens) [inline runtests.sh]: can we remove ` -not -path 'snapcraft/vendor/*'` ?

File: runtests.sh [hunk 7]
```diff
@@ -70,7 +65,7 @@ run_static_tests(){
 run_unit_tests(){
     if which python3-coverage >/dev/null 2>&1; then
         python3-coverage erase
-        python3-coverage run --branch --source snapcraft -m unittest discover -s snapcraft -t .
+        python3-coverage run --branch --source snapcraft --omit 'snapcraft/vendor/*' -m unittest discover -s snapcraft -t .
```
Comment (sergiusens) [inline runtests.sh]: can we remove `--omit 'snapcraft/vendor/*'` ?

File: snapcraft/commands/upload.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+def main(argv=None):
+    """Upload snap package to the Ubuntu Store."""
+
+    # make sure the full lifecycle is executed
+    yaml_config = snapcraft.yaml.load_config()
+    snap_name = _format_snap_name(yaml_config.data)
+
+    if not os.path.exists(snap_name):
+        logger.info(
+            'Snap {} not found. Running snap stage to create it.'.format(
```
Comment (sergiusens) [inline snapcraft/commands/upload.py]: `s/snap stage/snap step` please

File: snapcraft/commands/upload.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+def _format_snap_name(snap):
+    snap['arch'] = (snap['architectures'][0]
+                    if len(snap['architectures']) == 1 else 'multi')
+    return '{name}_{version}_{arch}.snap'.format(**snap)
+
+
+def main(argv=None):
+    """Upload snap package to the Ubuntu Store."""
+
```
Comment (sergiusens) [inline snapcraft/commands/upload.py]: For help to work, we need this here:

```python
argv = argv if argv else []
args = docopt(__doc__, argv=argv)
```

File: snapcraft/commands/logout.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+from snapcraft.config import clear_config
+
+
+logger = logging.getLogger(__name__)
+
+
+def main(argv=None):
+    """Forget credentials for Ubuntu One SSO."""
+
```
Comment (sergiusens) [inline snapcraft/commands/logout.py]: For help to work, we need this here:

```python
argv = argv if argv else []
args = docopt(__doc__, argv=argv)
```

File: snapcraft/commands/login.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+from snapcraft.config import save_config
+from snapcraft.storeapi import login
+
+
+logger = logging.getLogger(__name__)
+
+
+def main(argv=None):
+    """Authenticates session against Ubuntu One SSO."""
```
Comment (sergiusens) [inline snapcraft/commands/login.py]: For help to work, we need this here:

```python
argv = argv if argv else []
args = docopt(__doc__, argv=argv)
```

**Comment by sergiusens** [comment]
Looks pretty good, the only thing I think is missing is some sort of upload progress

```
$ ../../bin/snapcraft upload --help
Uploading shout_0.52.0_amd64.snap
Uploading files...
Starting new HTTPS connection (1): upload.apps.ubuntu.com
Uploading new version...
Starting new HTTPS connection (1): myapps.developer.ubuntu.com
Package submitted to https://myapps.developer.ubuntu.com/dev/api/click-package-upload/shout/
Checking package status...
... retrying in 5 seconds
Resetting dropped connection: myapps.developer.ubuntu.com
Package scan completed.
Upload did not complete.
Some errors were detected:

The uploaded version (0.52.0) has already been uploaded.


shout_0.52.0_amd64.snap upload failed
```

Thanks for working on this btw, we really appreciate the contribution!!!!

**Comment by sergiusens** [comment]
I'm sorry to say the error seen here makes little sense to me, it might just be a network error, in any case, mind doing a:

```
git checkout master
git pull
git checkout <this-branch>
git rebase master
# also add the lp bug this is fixing (add "LP: #1538657" on a single line)
git commit --amend
git push --force <your-origin>
```

**Comment by sergiusens** [comment]
OK, I've decided to let this in as is, thanks.

I've logged https://bugs.launchpad.net/snapcraft/+bug/1538692

**Pull request merged**